agnpy/time_evolution/_time_evolution_utils.py

```diff
@@ -0,0 +1,238 @@
+    # Convert each to list, adding first element
+    return [[g[0]-1, *list(g)] for g in groups]
+
+
+def merge_points(x, y, merge_groups, interpolated_distribution):
+    """
+    Merge groups of points in (x, y).
+    Ideally we should try to make it such that integral is preserved, but in special cases this could lead to
+    negative densities. So instead we just take the average.
+
+    Parameters:
+    - x: 1D numpy array, x-coordinates (must be sorted)
+    - y: 1D numpy array, y-coordinates
+    - merge_groups: list of lists, each inner list contains indices of x to merge
+
+    Returns:
+    - x_new, y_new: arrays with merged points
+    """
+
+    labels = np.full(len(x), -1)
+    for g_idx, group in enumerate(merge_groups):
+        labels[group] = g_idx
+    x_means = np.array([np.exp(np.mean(np.log(x[labels == g]))) for g in range(len(merge_groups))])
+    if interpolated_distribution is None:
+        y_means = Quantity([np.mean(y[labels == g]) for g in range(len(merge_groups))])
+    else:
+        y_means = Quantity([interpolated_distribution(x_means[g]) for g in range(len(merge_groups))])
+    group_tails = [item for group in merge_groups for item in group[1:]]
+    retain_mask = ~np.isin(np.arange(len(x)), group_tails)
+    labels_destination = labels[retain_mask]
+    merged_x = np.where(labels_destination == -1, x[retain_mask], x_means[labels_destination])
+    merged_y = np.where(labels_destination == -1, y[retain_mask], y_means[labels_destination])
+    return merged_x, merged_y
+
+
+def remove_empty_densities(result: TimeEvaluationResult) -> TimeEvaluationResult:
+    mask = result.density > 0
+    return TimeEvaluationResult(
+        result.total_time,
+        result.gamma[mask],
+        result.density[mask],
+        remap_subgroups_density(mask_to_mapping(mask), result.density_subgroups),
+        remap(mask, result.en_chg_rates),
+        remap(mask, result.abs_inj_rates),
+        remap(mask, result.rel_inj_rates))
+
+
+def update_distribution(gamma_array, n_array, blob):
+    gamma_sorted_and_merged, n_array_sorted_and_merged, _ = sort_and_merge_duplicates(gamma_array, n_array, None)
+    if np.all(n_array_sorted_and_merged == 0):
+        distribution = EmptyDistribution(gamma_min=gamma_sorted_and_merged[0], gamma_max=gamma_sorted_and_merged[-1])
+    else:
+        if len(gamma_sorted_and_merged[0]) == 1:
+            raise DistributionToSinglePointCollapseError(gamma_sorted_and_merged[0])
+        distribution = InterpolatedDistribution(gamma_sorted_and_merged[0], n_array_sorted_and_merged,
+                                     interpolator=PchipInterpolator)
+    blob.n_e = distribution
+    return distribution
+
+
+def energy_changes_lb(energy_changes_dict) -> dict:
+    return {key: value[0] for key, value in energy_changes_dict.items()}
+
+
+def to_total_energy_gev_sqr(gm_bins, density, total_volume):
+    ev_per_gamma = (m_e * c ** 2).to("eV")
+    gev_per_gamma = ev_per_gamma * 1e-9
```

Comment:
gev_per_gamma will still have a unit of eV, but value 10^-9 smaller - this is very confusing and will also affect the units in the next lines

Reply:
fixed